### PR TITLE
Parallel incremental checkpointing

### PR DESCRIPTION
**Pull request by SergeyTeplyakov**

### CODE REVIEW

File: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs [hunk 1]
```diff
@@ -128,37 +131,16 @@ private async Task CreateFullCheckpointAsync(OperationContext context, EventSequ
...
 
-        private async Task<bool> CreateCheckpointIncrementalAsync(OperationContext context, EventSequencePoint sequencePoint, bool successfullyUpdatedIncrementalState)
+        private async Task CreateCheckpointIncrementalAsync(OperationContext context, EventSequencePoint sequencePoint)
         {
             InitializeIncrementalCheckpointIfNeeded(restoring: false);
 
-            BoolResult result = BoolResult.Success;
             var incrementalCheckpointsPrefix = $"incrementalCheckpoints/{sequencePoint.SequenceNumber}.{Guid.NewGuid()}.";
-            var newCheckpointInfo = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            var newCheckpointInfo = new ConcurrentDictionary<string, string>(StringComparer.OrdinalIgnoreCase);
```
Comment (dcwuser) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs]: Comment describing meaning of keys and values.

File: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs [hunk 2]
```diff
@@ -170,11 +152,55 @@ private async Task<bool> CreateCheckpointIncrementalAsync(OperationContext conte
 
             await _checkpointRegistry.RegisterCheckpointAsync(context, checkpointId, sequencePoint).ThrowIfFailure();
             UpdateIncrementalCheckpointInfo(newCheckpointInfo);
-            successfullyUpdatedIncrementalState = true;
-            return successfullyUpdatedIncrementalState;
         }
 
-        private void WriteCheckpointInfo(AbsolutePath path, Dictionary<string, string> newCheckpointInfo)
+        private async Task UploadFilesAsync(OperationContext context, List<AbsolutePath> files, ConcurrentDictionary<string, string> newCheckpointInfo, string incrementalCheckpointsPrefix)
```
Comment (dcwuser) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs]: Can be just IEnumerable<AbsolutePath>, which also means we can avoid ToList conversion above.

File: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs [hunk 3]
```diff
@@ -128,37 +131,16 @@ private async Task CreateFullCheckpointAsync(OperationContext context, EventSequ
             await _checkpointRegistry.RegisterCheckpointAsync(context, checkpointId, sequencePoint).ThrowIfFailure();
         }
 
-        private async Task<bool> CreateCheckpointIncrementalAsync(OperationContext context, EventSequencePoint sequencePoint, bool successfullyUpdatedIncrementalState)
+        private async Task CreateCheckpointIncrementalAsync(OperationContext context, EventSequencePoint sequencePoint)
```
Comment (dcwuser) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs]: Logging with timings of steps?

File: Public/Src/Cache/ContentStore/Distributed/NuCache/DistributedCentralStorage.cs [hunk 4]
```diff
@@ -108,13 +108,15 @@ protected override async Task<BoolResult> ShutdownCoreAsync(OperationContext con
...
 
             // Need to touch in fallback storage as well so it knows the content is still in use
-            await _fallbackStorage.TouchBlobAsync(context, file, fallbackStorageId, isUploader).ThrowIfFailure();
+            var touchTask = _fallbackStorage.TouchBlobAsync(context, file, fallbackStorageId, isUploader).ThrowIfFailure();
 
             // Ensure content is present in private CAS and registered
-            await PutAndRegisterFileAsync(context, file, hash);
+            var registerTask = PutAndRegisterFileAsync(context, file, hash);
+
+            await Task.WhenAll(touchTask, registerTask);
```
Comment (dcwuser) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/DistributedCentralStorage.cs]: Unread exceptions?

File: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs (hunk 3 above)
Comment (SergeyTeplyakov) [reply Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs]: Each step is already wrapped in it's own performoperation.
So I don't think we need to do this explicitely once again.

---
In reply to: [319297670](https://github.com/microsoft/BuildXL/pull/818#discussion_r319297670) [](ancestors = 319297670)

File: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs (hunk 1 above)
Comment (SergeyTeplyakov) [reply Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs]: Added a comment to see the comment of _incrementalCheckpointInfo field.

---
In reply to: [319296827](https://github.com/microsoft/BuildXL/pull/818#discussion_r319296827) [](ancestors = 319296827)

File: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs (hunk 2 above)
Comment (SergeyTeplyakov) [reply Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs]: I actually would slightly prefer to have ToList on the source to separate enumeration process from the consumption process...

---
In reply to: [319297335](https://github.com/microsoft/BuildXL/pull/818#discussion_r319297335) [](ancestors = 319297335)

File: Public/Src/Cache/ContentStore/Distributed/NuCache/DistributedCentralStorage.cs (hunk 4 above)
Comment (SergeyTeplyakov) [reply Public/Src/Cache/ContentStore/Distributed/NuCache/DistributedCentralStorage.cs]: As we discussed, WhenAll throws an error.

---
In reply to: [319297916](https://github.com/microsoft/BuildXL/pull/818#discussion_r319297916) [](ancestors = 319297916)

File: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs [hunk 5]
```diff
@@ -303,7 +336,14 @@ public Task<BoolResult> RestoreCheckpointAsync(OperationContext context, string
 
                             if (isIncrementalCheckpoint)
                             {
-                                successfullyUpdatedIncrementalState = await RestoreCheckpointIncrementalAsync(context, checkpointFile, extractedCheckpointDirectory);
+                                var restoreCheckpointResult = await RestoreCheckpointIncrementalAsync(context, checkpointFile, extractedCheckpointDirectory);
+                                successfullyUpdatedIncrementalState = restoreCheckpointResult;
+
+                                if (!restoreCheckpointResult)
+                                {
+                                    _tracer.Debug(context, $"Failed restoring incremental checkpoint '{restoreCheckpointResult}'. Restoring the full checkpoing...");
```
Comment (ljcollins25) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs]: >checkpoing [](start = 149, length = 10)

checkpoint

**Pull request merged**